File: packages/indexer/src/extract.ts
import { detectLanguage, getRegisteredLanguageCodes } from "@csf/analysis";
import { parse } from "node-html-parser";

export interface PinDeclaration {
  phrase: string;
  mode: "exact" | "contains";
  priority: number;
  exclusive: boolean;
}

export interface ExtractedDocument {
  title: string;
  language: string;
  body: string;
  excerpt: string;
  url: string;
  noindex: boolean;
  boost: number;
  /** Facet field name -> distinct values declared via csf-facet-<field> (docs/15-cms-meta-tag-control.md). */
  facets: Record<string, string[]>;
  /** Range-facet field name -> single numeric value declared via csf-facet-range-<field> (docs/06-faceted-search.md). One value per doc, unlike terms facets. */
  rangeFacets: Record<string, number>;
  pins: PinDeclaration[];
}

const FACET_TAG_PREFIX = "csf-facet-";
const RANGE_FACET_TAG_PREFIX = "csf-facet-range-";

export interface CanonicalUrlOptions {
  /**
   * Restricts an absolute canonical URL to one of these exact origins
   * (scheme + host + port, e.g. `"https://example.com"`) — anything
   * else falls back to the document's own crawled `sourceUrl`, same as
   * a missing/invalid canonical tag. Off by default: with no allowlist
   * configured, any `http:`/`https:` origin is accepted (see
   * `SAFE_URL_PROTOCOLS` below) since most deployments have no fixed
   * production origin at index time (e.g. a preview/staging build).
   */
  allowedUrlOrigins?: string[];
  /**
   * Base URL used to resolve a root-relative (`"/products/widget"`) or
   * otherwise relative canonical href into an absolute URL before
   * protocol/origin checks run. Without this, only already-absolute
   * `http:`/`https:` canonical URLs and bare root-relative paths are
   * accepted — a canonical href that's relative to the *current* page
   * (not root-relative) has no way to resolve to anything meaningful
   * without knowing that page's own URL, so it's rejected (falls back
   * to `sourceUrl`) rather than guessed at.
   */
  baseUrl?: string;
}

/**
 * Schemes a canonical URL is allowed to resolve to (docs/15-cms-meta-tag-control.md#canonical-url):
 * `javascript:`, `data:`, and other executable/non-web schemes are
 * rejected outright, since this value flows through to `Hit.url`, which
 * a consuming app may render directly into a link — untrusted/malformed
 * rendered HTML producing one of those schemes would otherwise be a
 * real XSS footgun handed straight to every consumer.
 */
const SAFE_URL_PROTOCOLS = new Set(["http:", "https:"]);

/**
 * Validates a `<link rel="canonical">` href before trusting it as a
 * document's `url` (docs/15-cms-meta-tag-control.md#canonical-url) —
 * malformed, non-web-scheme, or (when `allowedUrlOrigins` is set)
 * off-allowlist values fall back to `sourceUrl`, exactly as if no
 * canonical tag were present at all, rather than either throwing (one
 * bad page souring an entire build) or passing the raw value through
 * unchecked (the security footgun this exists to close).
 */
function sanitizeCanonicalUrl(
  canonical: string | undefined,
  sourceUrl: string,
  options: CanonicalUrlOptions,
): string {
  if (!canonical) return sourceUrl;

  // Root-relative paths ("/products/widget") have no scheme to exploit
  // and need no base to resolve -- accepted as-is, matching how a real
  // browser treats <link rel="canonical" href="/foo">. Protocol-relative
  // ("//evil.com/foo") is deliberately excluded from this fast path (its
  // second character is also "/") so it always goes through full URL
  // resolution below instead of being trusted blind.
  if (canonical.startsWith("/") && !canonical.startsWith("//")) {
    return canonical;
  }

  let parsed: URL;
  try {
    parsed = new URL(canonical, options.baseUrl);
  } catch {
    console.warn(
      `[csf-indexer] canonical URL "${canonical}" for ${sourceUrl} is malformed or relative with no baseUrl configured -- ignoring, indexing with ${sourceUrl} instead. See docs/15-cms-meta-tag-control.md#canonical-url.`,
    );
    return sourceUrl;
  }

  if (!SAFE_URL_PROTOCOLS.has(parsed.protocol)) {
    console.warn(
      `[csf-indexer] canonical URL "${canonical}" for ${sourceUrl} uses a disallowed protocol (${parsed.protocol}) -- ignoring, indexing with ${sourceUrl} instead.`,
    );
    return sourceUrl;
  }

  if (
    options.allowedUrlOrigins &&
    !options.allowedUrlOrigins.includes(parsed.origin)
  ) {
    console.warn(
      `[csf-indexer] canonical URL "${canonical}" for ${sourceUrl} resolves to an origin (${parsed.origin}) not in allowedUrlOrigins -- ignoring, indexing with ${sourceUrl} instead.`,
    );
    return sourceUrl;
  }

  // `parsed.href`, not the raw `canonical` string: this is the point at
  // which a protocol-relative href (accepted here only because `baseUrl`
  // resolved it to a real origin) needs to become the actual absolute
  // URL, not the still-schemeless original.
  return parsed.href;
}

const BOILERPLATE_SELECTORS = [
  "nav",
  "header",
  "footer",
  "aside",
  "script",
  "style",
];

function collapseWhitespace(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

/**
 * Extracts indexable fields from one rendered HTML page, honoring the
 * csf-* meta-tag control surface (docs/15-cms-meta-tag-control.md):
 * title/body/language/excerpt/canonical/noindex/boost plus facet
 * values (csf-facet-<field>) and pin declarations (csf-pin*, see
 * docs/16-term-to-page-pinning.md).
 */
export function extractDocument(
  html: string,
  sourceUrl: string,
  defaultLanguage = "en",
  canonicalUrlOptions: CanonicalUrlOptions = {},
): ExtractedDocument {
  const root = parse(html);

  const noindex = root.querySelector('meta[name="csf-noindex"]') !== null;

  const title = collapseWhitespace(
    root.querySelector("title")?.structuredText ?? "",
  );

  const declaredLanguage = root
    .querySelector("html")
    ?.getAttribute("lang")
    ?.trim();

  const canonical = root
    .querySelector('link[rel="canonical"]')
    ?.getAttribute("href")
    ?.trim();
  const url = sanitizeCanonicalUrl(canonical, sourceUrl, canonicalUrlOptions);

  const excerpt = collapseWhitespace(
    root.querySelector('meta[name="description"]')?.getAttribute("content") ??
      "",
  );

  const bodyRoot =
    root.querySelector("[data-csf-body]") ??
    root.querySelector("main") ??
    root.querySelector("body");

  for (const el of bodyRoot?.querySelectorAll(
    [...BOILERPLATE_SELECTORS, "[data-csf-ignore]"].join(","),
  ) ?? []) {
    el.remove();
  }

  const body = collapseWhitespace(bodyRoot?.structuredText ?? "");

  // Explicit <html lang> always wins; detectLanguage() is only a
  // fallback for pages that declare none (docs/03-tokenization-i18n.md,
  // docs/09-roadmap.md's "Auto language detection accuracy" open
  // question) -- a deterministic, zero-bundled-model heuristic, not a
  // substitute for authoring real language metadata. Falls back to
  // defaultLanguage when detection itself has no confident signal
  // (e.g. too little text, or a language with no detection heuristic),
  // exactly the prior behavior for an undeclared page.
  const language =
    declaredLanguage ||
    detectLanguage(`${title} ${body}`, getRegisteredLanguageCodes()) ||
    defaultLanguage;

  const boostAttr = root
    .querySelector('meta[name="csf-boost"]')
    ?.getAttribute("content");
  const parsedBoost = boostAttr ? Number.parseFloat(boostAttr) : Number.NaN;
  const boost =
    Number.isFinite(parsedBoost) && parsedBoost > 0 ? parsedBoost : 1.0;

  const facets: Record<string, string[]> = {};
  const rangeFacets: Record<string, number> = {};
  for (const meta of root.querySelectorAll("meta")) {
    const name = meta.getAttribute("name") ?? "";
    // Checked first: csf-facet-range-<field> also starts with the plain
    // csf-facet- prefix below, so it would otherwise be misparsed as a
    // terms facet field literally named "range-<field>".
    if (name.startsWith(RANGE_FACET_TAG_PREFIX)) {
      const field = name.slice(RANGE_FACET_TAG_PREFIX.length);
      const raw = meta.getAttribute("content")?.trim();
      const parsed = raw ? Number.parseFloat(raw) : Number.NaN;
      // First declared value wins; a page authoring the same range
      // field twice is almost certainly a mistake, not intentional.
      if (field && Number.isFinite(parsed) && !(field in rangeFacets)) {
        rangeFacets[field] = parsed;
      }
      continue;
    }
    if (!name.startsWith(FACET_TAG_PREFIX)) continue;
    const field = name.slice(FACET_TAG_PREFIX.length);
    const value = meta.getAttribute("content")?.trim();
    if (!field || !value) continue;
    const values = facets[field] ?? [];
    if (!values.includes(value)) values.push(value);
    facets[field] = values;
  }

  const pinPhrases = root
    .querySelectorAll('meta[name="csf-pin"]')
    .map((el) => el.getAttribute("content")?.trim())
    .filter((v): v is string => Boolean(v));

  const pinModeAttr = root
    .querySelector('meta[name="csf-pin-mode"]')
    ?.getAttribute("content")
    ?.trim();
  const pinMode: "exact" | "contains" =
    pinModeAttr === "contains" ? "contains" : "exact";

  const pinPriorityAttr = root
    .querySelector('meta[name="csf-pin-priority"]')
    ?.getAttribute("content");
  const parsedPriority = pinPriorityAttr
    ? Number.parseFloat(pinPriorityAttr)
    : Number.NaN;
  const pinPriority = Number.isFinite(parsedPriority) ? parsedPriority : 0;

  // Presence-based, like csf-noindex above: the tag's content isn't
  // interpreted, only whether the tag exists on the page at all.
  const pinExclusive =
    root.querySelector('meta[name="csf-pin-exclusive"]') !== null;

  const pins: PinDeclaration[] = pinPhrases.map((phrase) => ({
    phrase,
    mode: pinMode,
    priority: pinPriority,
    exclusive: pinExclusive,
  }));

  return {
    title,
    language,
    body,
    excerpt,
    url,
    noindex,
    boost,
    facets,
    rangeFacets,
    pins,
  };
}

File: python/searchable-indexer/src/searchable_indexer/extract.py
import math
import re
import sys
from urllib.parse import urljoin, urlparse

from selectolax.parser import HTMLParser

from searchable_analysis import detect_language, get_registered_language_codes
from searchable_indexer.types import ExtractedDocument, PinDeclaration

_FACET_TAG_PREFIX = "searchable-facet-"
_RANGE_FACET_TAG_PREFIX = "searchable-facet-range-"
_BOILERPLATE_SELECTORS = ["nav", "header", "footer", "aside", "script", "style"]
_SAFE_URL_PROTOCOLS = {"http", "https"}
_WHITESPACE_RE = re.compile(r"\s+")


def _collapse_whitespace(text: str) -> str:
    return _WHITESPACE_RE.sub(" ", text).strip()


def _parse_float_or_nan(raw: str | None) -> float:
    if not raw:
        return float("nan")
    try:
        return float(raw)
    except ValueError:
        return float("nan")


def _warn(message: str) -> None:
    print(f"[searchable-indexer] {message}", file=sys.stderr)


def _sanitize_canonical_url(
    canonical: str,
    source_url: str,
    allowed_url_origins: list[str] | None,
    base_url: str | None,
) -> str:
    if not canonical:
        return source_url

    # Root-relative paths have no scheme to exploit and need no base to
    # resolve. Protocol-relative ("//evil.com/foo") is deliberately
    # excluded from this fast path (its second character is also "/")
    # so it always goes through full URL resolution below.
    if canonical.startswith("/") and not canonical.startswith("//"):
        return canonical

    direct = urlparse(canonical)
    if direct.scheme and direct.netloc:
        parsed = direct
    elif base_url:
        parsed = urlparse(urljoin(base_url, canonical))
    else:
        _warn(
            f'canonical URL "{canonical}" for {source_url} is malformed or relative '
            "with no baseUrl configured -- ignoring, indexing with "
            f"{source_url} instead."
        )
        return source_url

    if parsed.scheme not in _SAFE_URL_PROTOCOLS or not parsed.netloc:
        _warn(
            f'canonical URL "{canonical}" for {source_url} uses a disallowed '
            f"protocol ({parsed.scheme}:) -- ignoring, indexing with {source_url} instead."
        )
        return source_url

    if allowed_url_origins is not None:
        origin = f"{parsed.scheme}://{parsed.hostname}"
        if parsed.port:
            origin += f":{parsed.port}"
        if origin not in allowed_url_origins:
            _warn(
                f'canonical URL "{canonical}" for {source_url} resolves to an '
                f"origin ({origin}) not in allowed_url_origins -- ignoring, "
                f"indexing with {source_url} instead."
            )
            return source_url

    return parsed.geturl()


def extract_document(
    html: str,
    source_url: str,
    default_language: str = "en",
    allowed_url_origins: list[str] | None = None,
    canonical_base_url: str | None = None,
) -> ExtractedDocument:
    tree = HTMLParser(html)

    noindex = tree.css_first('meta[name="searchable-noindex"]') is not None

    title_node = tree.css_first("title")
    title = _collapse_whitespace(title_node.text(deep=True, separator=" ") if title_node else "")

    html_node = tree.css_first("html")
    declared_language = (
        (html_node.attributes.get("lang") or "").strip() if html_node else ""
    )

    canonical_node = tree.css_first('link[rel="canonical"]')
    canonical = (
        (canonical_node.attributes.get("href") or "").strip() if canonical_node else ""
    )
    url = _sanitize_canonical_url(
        canonical, source_url, allowed_url_origins, canonical_base_url
    )

    desc_node = tree.css_first('meta[name="description"]')
    excerpt = _collapse_whitespace(
        (desc_node.attributes.get("content") or "") if desc_node else ""
    )

    body_node = (
        tree.css_first("[data-searchable-body]")
        or tree.css_first("main")
        or tree.css_first("body")
    )
    if body_node is not None:
        selector = ",".join([*_BOILERPLATE_SELECTORS, "[data-searchable-ignore]"])
        for el in body_node.css(selector):
            el.decompose()
    body = _collapse_whitespace(body_node.text(deep=True, separator=" ") if body_node else "")

    language = (
        declared_language
        or detect_language(f"{title} {body}", get_registered_language_codes())
        or default_language
    )

    boost_node = tree.css_first('meta[name="searchable-boost"]')
    parsed_boost = _parse_float_or_nan(
        boost_node.attributes.get("content") if boost_node else None
    )
    boost = parsed_boost if math.isfinite(parsed_boost) and parsed_boost > 0 else 1.0

    facets: dict[str, list[str]] = {}
    range_facets: dict[str, float] = {}
    for meta in tree.css("meta"):
        name = meta.attributes.get("name") or ""
        if name.startswith(_RANGE_FACET_TAG_PREFIX):
            field_name = name[len(_RANGE_FACET_TAG_PREFIX) :]
            raw = (meta.attributes.get("content") or "").strip()
            parsed = _parse_float_or_nan(raw)
            if field_name and math.isfinite(parsed) and field_name not in range_facets:
                range_facets[field_name] = parsed
            continue
        if not name.startswith(_FACET_TAG_PREFIX):
            continue
        field_name = name[len(_FACET_TAG_PREFIX) :]
        value = (meta.attributes.get("content") or "").strip()
        if not field_name or not value:
            continue
        values = facets.setdefault(field_name, [])
        if value not in values:
            values.append(value)

    pin_phrases = [
        (el.attributes.get("content") or "").strip()
        for el in tree.css('meta[name="searchable-pin"]')
    ]
    pin_phrases = [p for p in pin_phrases if p]

    pin_mode_node = tree.css_first('meta[name="searchable-pin-mode"]')
    pin_mode_attr = (
        (pin_mode_node.attributes.get("content") or "").strip() if pin_mode_node else ""
    )
    pin_mode = "contains" if pin_mode_attr == "contains" else "exact"

    pin_priority_node = tree.css_first('meta[name="searchable-pin-priority"]')
    parsed_priority = _parse_float_or_nan(
        pin_priority_node.attributes.get("content") if pin_priority_node else None
    )
    pin_priority = parsed_priority if math.isfinite(parsed_priority) else 0.0

    pin_exclusive = tree.css_first('meta[name="searchable-pin-exclusive"]') is not None

    pins = [
        PinDeclaration(
            phrase=phrase, mode=pin_mode, priority=pin_priority, exclusive=pin_exclusive
        )
        for phrase in pin_phrases
    ]

    return ExtractedDocument(
        title=title,
        language=language,
        body=body,
        excerpt=excerpt,
        url=url,
        noindex=noindex,
        boost=boost,
        facets=facets,
        range_facets=range_facets,
        pins=pins,
    )
